quantum/util/impl/quantum_sequencer_lite_impl.h

```diff
@@ -0,0 +1,475 @@
+bool
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::addPendingTask(
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    bool empty = _universalTaskQueue._tasks.empty();
+    _universalTaskQueue._tasks.push_back(task);
+    if (not empty)
+    {
+        ++task->_pendingKeyCount;
+    }
+    _universalTaskQueue._stats->incrementPostedTaskCount();
+    _universalTaskQueue._stats->incrementPendingTaskCount();
+
+    return empty;
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+bool
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::addPendingTask(
+    const SequenceKey& key,
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    typename PendingTaskQueueMap::iterator it = _pendingTaskQueueMap.find(key);
+    if (it == _pendingTaskQueueMap.end())
+    {
+        it = _pendingTaskQueueMap.emplace(key, SequencerLiteKeyData<SequenceKey>()).first;
+        // add each universal task to this queue before the task
+        for(auto& universalTask : _universalTaskQueue._tasks)
+        {
+            it->second._tasks.push_back(universalTask);
+            ++universalTask->_pendingKeyCount;
+        }
+    }
+    bool empty = it->second._tasks.empty();
+    it->second._tasks.push_back(task);
+    if (not empty)
+    {
+        ++task->_pendingKeyCount;
+    }
+    if (not task->_universal)
+    {
+        task->_keyData.push_back(&it->second);
+        it->second._stats->incrementPostedTaskCount();
+        it->second._stats->incrementPendingTaskCount();
+    }
+    return empty;
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+std::shared_ptr<SequencerLiteTask<SequenceKey>>
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::removePending(
+    SequencerLiteKeyData<SequenceKey>& entry,
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    // delete this entry
+    if (entry._tasks.empty() or
+        entry._tasks.front() != task)
+    {
+        return std::shared_ptr<SequencerLiteTask<SequenceKey>>();
+    }
+
+    entry._tasks.pop_front();
+    if (entry._tasks.empty())
+    {
+        return std::shared_ptr<SequencerLiteTask<SequenceKey>>();
+    }
+    std::shared_ptr<SequencerLiteTask<SequenceKey>> next = entry._tasks.front();
+    // the new front has one less key to wait
+    return 0 == --next->_pendingKeyCount
+        ? next
+        : std::shared_ptr<SequencerLiteTask<SequenceKey>>();
+}
+
+template <class SequenceKey, class Hash, class KeyEqual, class Allocator>
+void
+SequencerLite<SequenceKey, Hash, KeyEqual, Allocator>::removePending(
+    const std::shared_ptr<SequencerLiteTask<SequenceKey>>& task)
+{
+    std::lock_guard<std::mutex> lock(_mutex);
+    if (task->_universal)
+    {
+        // remove the task from all key queues
+        for(auto& item : _pendingTaskQueueMap)
+        {
+            if (std::shared_ptr<SequencerLiteTask<SequenceKey>> nextTask = removePending(item.second, task))
```

Comment:
Let's use your example: we have two pending universal tasks T1 and T2 all enqueued into N queues + the universal queue. Then assume we start executing task T1, and once it's done, SequencerLite will call removePending for T1. Before the for-loop from line 128 starts, the T2's _pendingKeyCount will be equal to N (because T2 is not at the head of any of the N queues yet). Each call of removingPending for T1 will result in T2 becoming at the head of the corresponding queue and reducing T2->_pendingKeyCount  by one. However, those removePending will be always returning nullptr (see lines 114-116) until the final call that will make T2->_pendingKeyCount equal to 0 and hence returning it as the next task to execute. Hence T2 will be executed only once.

Reply:
Ah ok...good. I was not sure if this scenario was handled or not.